src/freenet/clients/http/ToadletContextImpl.java

```diff
@@ -376,15 +376,32 @@ static void sendReplyHeaders(OutputStream sockOutputStream, int replyCode, Strin
 			}
 		if(contentLength >= 0)
 			mvt.put("content-length", Long.toString(contentLength));
-		
-		String expiresTime;
+
+		boolean allowCaching; // For privacy reasons, only static
+							  // content may be cached
 		if (mTime == null) {
-			expiresTime = "Thu, 01 Jan 1970 00:00:00 GMT";
+			allowCaching = false;
+		} else {
+			allowCaching = true;
+		}
+		String expiresTime;
+		String cacheControl;
+		if (allowCaching) {
+			// use an expiry time of 30 day from now, about the frequency of Freenet releases
+			// Expires is needed for older browsers
+			expiresTime = TimeUtil.makeHTTPDate(System.currentTimeMillis() + DAYS.toMillis(30));
+			cacheControl = "public, max-age=" + String.valueOf(3600 * 24 * 30);
+			// Use contentlength as simplest possible hash function
+			// which needs no additional file access
+			mvt.put("ETag", String.valueOf(contentLength));
```

Comment:
There are virtually unlimited changes you can perform on a file without changing its length so the length of a file is pretty much the worst hash value you could ever have (except for a constant value).

What we need here is a value that does change if the file changes; servlet start time would be an option, build version is slightly worse (because of pre-releases which have the same build version as other builds and the files might have changed), even the real hash would be okay because we only need to calculate it once; static assets can additionally be cached because they don’t change at runtime: if they have already been requested once during Fred’s lifetime they are guaranteed to not have changed since then.